c10/util/BFloat16.h

```diff
@@ -0,0 +1,50 @@
+#pragma once
+
+// Defines the bloat16 type (brain floating-point). This representation uses
+// 1 bit for the sign, 8 bits for the exponent and 7 bits for the mantissa.
+
+#include <c10/macros/Macros.h>
+#include <cmath>
+#include <cstring>
+
+namespace c10 {
+
+namespace detail {
+  inline C10_HOST_DEVICE float f32_from_bits(uint16_t src) {
+    float* res = 0;
+    uint32_t tmp = src;
+    tmp <<= 16;
+
+    // We should be using memcpy in order to respect the strict aliasing rule
+    // but it fails in the HIP environment.
+    res = reinterpret_cast<float*>(&tmp);
```

Comment:
I think it's better if you `ifdef` here and do the correct thing in non-HIP and the incorrect thing in HIP.  And then we can just get rid of the ifdef once HIP is fixed and we'll already have the correct code written.  As you have it now, you'll have to remember what you did when that time comes.